### ISSUE TITLE
🚀 Enable gradient pass through all biquad filter function

### ISSUE THREAD

**Issue by yoyololicon** [pytorch/audio#1375]
## 🚀 Feature

I propose make all `*_biquad` filter function become differentiable respect to their parameters (`Q`, `gain`, `central_freq`... etc.)

## Motivation

Recently there is a trend on using differentiable IIR filters to learn frequency responses efficiently. 
In the last year DAFX2020 already about 3 papers described similar idea. Some are produced by well known companies (Native Instruments, iZotope). Some also released their [source code](https://github.com/boris-kuz/differentiable_iir_filters), but to enable automatic differentiation, they use an RNN-like, for-loop structure, which is relatively slow and memory inefficient. 
Because our implementation of IIR (the `lfilter`) use in-place operation, it opens up possibilities to offer fast and accurate differentiable IIR filters.

## Pitch

Change parameters (like `gain`, `Q`, `central_freq`) type hint from `float` to `Union[float, Tensor]`, and replace those `math.*` function to corressponding `torch.*` one.

## Additional context

For references:

- Differentiable IIR Filters for Machine Learning Applications
Boris Kuznetsov; Julian Parker; Fabian Esqueda
- Optimization of Cascaded Parametric Peak and Shelving Filters With Backpropagation Algorithm
Purbaditya Bhattacharya; Patrick Nowak; Udo Zölzer
- Neural Parametric Equalizer Matching Using Differentiable Biquads
Shahan Nercessian

**Comment by mthrok** [comment]
Hi @yoyololicon 

Thanks for the proposal. We are also interested in differential DSP (like in (magenta)[https://magenta.tensorflow.org/ddsp]), but we did not get to work on it due to the lack of the engineering resource. I second your proposal, it sounds like a natural extension of the current `lfilter` update, it does not sound like involve BC-breaking change.

We will be needing tests for ensuring the autograd support for these filters, and we, torchaudio library, are new to the nature of autograd by ourselves so there might be some things that we need to sort out and make an official stance (like the standard for numerical precision mentioned in #1310, which is new for us and ideally we want to establish a good guideline for the tolerance.)

What do you think @cpuhrsch ?

**Comment by cpuhrsch** [comment]
This sounds great! I'm all for this.

**Comment by yoyololicon** [comment]
@cpuhrsch @mthrok 
I was prototyping the feature, and found that torch jit doesn't support union type like `Union[float, Tensor]`, leading to fail on torch script consistency tests.
Can we force users to pass tensor type on these parameters, or do you guys have other work around ideas that would not break backward supportability?

**Comment by mthrok** [comment]
> @cpuhrsch @mthrok
> I was prototyping the feature, and found that torch jit doesn't support union type like `Union[float, Tensor]`, leading to a fail on torch script consistency tests.
> Should we force users to pass tensor type on these parameters, or do you guys have other work around ideas that would not break backward capability?

Hi @yoyololicon

One workaround we have is to only use the annotation that is compatible with TorchScript. For example, I/O functions support file-like objects and path-like objects but we only annotate them as `str` and added note in docstring why we do so. https://pytorch.org/audio/stable/backend.html#info

However I am afraid this approach will not work completely for the case here. If we restrict the annotation of the function argument to one of `float` or `tensor`, once the function is scripted, it will lose the ability to handle the other. If it is not scripted, then we still should be able to handle the other case.

There is an ongoing effort on supporting `Union` type in TorchScript, https://github.com/pytorch/pytorch/issues/41412 and maybe this could be a non issue in the (near) future.

Meanwhile, I can think of three approaches. 

1. [BC-breaking] Change the annotation and expected type to `Tensor`. If float is passed, wrap it to Tensor.
    ```python
    def foo(value: Tensor):
        if isinsfance(value, float):
            value = torch.Tensor(float)
    ```
    The existing code will keep working as long as the function is not scripted.
    Once the function is scripted, users have to provide Tensor value, which can be annoying because one needs to wrap numeric literals with Tensor constructor.
    Note that we typically require 1 extra release cycle (3 ~ 4 months) when making BC-breaking changes to give users warning.
2. Keep the annotation as `float` but the function accepts `Tensor`.
    ```python
    def foo(value: float):
        if isinsfance(value, float):
            value = torch.Tensor(float)
    ```
    This approach will allow to add autograd support as long as the function is not scripted. (but the most common use case of TorchScript is for inference, and IIRC it is not straightforward to do autograd with TorchScript, so I think this is okay.)
3. Wait until https://github.com/pytorch/pytorch/issues/41412 is resolved.
    Once TorchScript supports `Union`, this will be no issue.

I think approach 2 and 3 can be combined. We can first add autograd support for pure-Python mode (called eager mode) without changing the interface, and add some notes explaining it in docstring. Then once TorchScript supports `Union`, then we can update the annotation and remove the note.

**Comment by yoyololicon** [comment]
@mthrok Great, let's go for (2).

**Comment by yoyololicon** [comment]
#1400 is now merged.
